File: shared/mcp-connectors/public/namespace_filter.py
#!/usr/bin/env python3
"""Namespace filter: which databases or schemas a connection may see (issue #31).

Dependency-free and kept under ``public/`` (the ``shared`` Docker build
context), like ``canonical_types.py``, so a connector image can ship it with
``COPY --from=shared namespace_filter.py /app/namespace_filter.py``.

The Go twin is ``backend-orchestrator/pkg/namespacefilter``; the connection
modal previews with a TypeScript copy,
``frontend/src/lib/pipeline/namespaceFilter.ts``. All three are pinned to the
same cases in ``namespace_filter_vectors.json`` (next to this file); change the
algorithm in all of them, or in none.

Config keys (connection config values are strings):
  * ``namespace_filter_mode``: ``all`` (default when missing or empty),
    ``include`` (only these) or ``exclude`` (all except). Trimmed and
    case-insensitive. Any other value is an error: never silently ``all``.
  * ``namespace_filter_patterns``: comma-separated. Entries are trimmed and
    empty entries dropped. At most 100 patterns, each at most 128 UTF-8 bytes.
    ``*`` matches any run of characters, including none. Every other character
    is literal. A pattern must match the whole name, case-insensitively.

``include`` with no patterns is an error. ``exclude`` with no patterns keeps
everything. System namespaces are always rejected first, whatever the mode.
Nothing matching is a warning for the caller, not an error: a database may
not exist yet.

Error and warning messages are fixed text with no names or patterns in them.
"""
from dataclasses import dataclass
from typing import Any, Iterable, List, Mapping, Tuple

MODE_KEY = "namespace_filter_mode"
PATTERNS_KEY = "namespace_filter_patterns"

MODE_ALL = "all"
MODE_INCLUDE = "include"
MODE_EXCLUDE = "exclude"
MODES = (MODE_ALL, MODE_INCLUDE, MODE_EXCLUDE)

MAX_PATTERNS = 100
MAX_PATTERN_BYTES = 128

NO_MATCH_WARNING = "The namespace filter matched no databases or schemas."


class NamespaceFilterError(ValueError):
    """The filter config is invalid. The caller must fail closed."""


@dataclass(frozen=True)
class NamespaceFilter:
    mode: str
    patterns: Tuple[str, ...]
    _lowered: Tuple[str, ...] = ()

    @property
    def active(self) -> bool:
        """True when the filter can drop a non-system name."""
        return self.mode == MODE_INCLUDE or (self.mode == MODE_EXCLUDE and bool(self.patterns))


@dataclass(frozen=True)
class ApplyResult:
    kept: List[str]
    matched: int
    excluded: int
    warning: str  # empty when there is nothing to warn about


def _glob_match(pattern: str, name: str) -> bool:
    """Whole-name glob match where ``*`` is the only wildcard.

    Iterative two-pointer matcher, O(len(pattern) * len(name)) worst case.
    A backtracking regex (``.*`` per star) is exponential in the star count
    for a non-matching name, so a user-authored pattern could hang discovery.
    Both arguments must already be lowercased.
    """
    p = n = 0
    star = -1
    mark = 0
    while n < len(name):
        if p < len(pattern) and pattern[p] != "*" and pattern[p] == name[n]:
            p += 1
            n += 1
        elif p < len(pattern) and pattern[p] == "*":
            star = p
            mark = n
            p += 1
        elif star != -1:
            p = star + 1
            mark += 1
            n = mark
        else:
            return False
    while p < len(pattern) and pattern[p] == "*":
        p += 1
    return p == len(pattern)


def parse(config: Mapping[str, Any]) -> NamespaceFilter:
    """Read the filter from a connection config. Raises NamespaceFilterError."""
    raw_mode = config.get(MODE_KEY) if config else None
    raw_patterns = config.get(PATTERNS_KEY) if config else None
    if raw_mode is None:
        raw_mode = ""
    if raw_patterns is None:
        raw_patterns = ""
    if not isinstance(raw_mode, str):
        raise NamespaceFilterError("namespace_filter_mode must be a string")
    if not isinstance(raw_patterns, str):
        raise NamespaceFilterError("namespace_filter_patterns must be a comma-separated string")

    mode = raw_mode.strip().lower() or MODE_ALL
    if mode not in MODES:
        raise NamespaceFilterError("namespace_filter_mode must be one of: all, include, exclude")

    patterns = tuple(p.strip() for p in raw_patterns.split(",") if p.strip())
    if len(patterns) > MAX_PATTERNS:
        raise NamespaceFilterError(f"namespace_filter_patterns allows at most {MAX_PATTERNS} patterns")
    for p in patterns:
        if len(p.encode("utf-8", "surrogatepass")) > MAX_PATTERN_BYTES:
            raise NamespaceFilterError(
                f"each namespace_filter_patterns entry must be at most {MAX_PATTERN_BYTES} bytes"
            )
    if mode == MODE_INCLUDE and not patterns:
        raise NamespaceFilterError("namespace_filter_mode 'include' needs at least one pattern")

    return NamespaceFilter(mode=mode, patterns=patterns, _lowered=tuple(p.lower() for p in patterns))


def _matches_any(name: str, f: NamespaceFilter) -> bool:
    lowered = name.lower()
    return any(_glob_match(p, lowered) for p in f._lowered)


def allowed(name: str, f: NamespaceFilter, system_names: Iterable[str] = ()) -> bool:
    """True when ``name`` passes: not a system namespace, then the mode applies."""
    lowered = name.lower()
    if any(lowered == s.lower() for s in system_names):
        return False
    if f.mode == MODE_INCLUDE:
        return _matches_any(name, f)
    if f.mode == MODE_EXCLUDE:
        return not _matches_any(name, f)
    return True


def apply(names: Iterable[str], f: NamespaceFilter, system_names: Iterable[str] = ()) -> ApplyResult:
    """Keep the allowed names in their input order.

    ``warning`` is set when an active filter leaves nothing, so the caller can
    tell the user instead of failing.
    """
    systems = list(system_names)
    candidates = list(names)
    kept = [n for n in candidates if allowed(n, f, systems)]
    warning = NO_MATCH_WARNING if f.active and not kept else ""
    return ApplyResult(kept=kept, matched=len(kept), excluded=len(candidates) - len(kept), warning=warning)

File: frontend/src/lib/pipeline/namespaceFilter.ts
/**
 * Namespace filter: which databases or schemas a connection may see (#31) —
 * the connection modal's Scope step. The server applies it (Python
 * shared/mcp-connectors/public/namespace_filter.py and Go
 * backend-orchestrator/pkg/namespacefilter); this copy only previews what a
 * pattern keeps before the connection is saved. All three are pinned to
 * shared/mcp-connectors/public/namespace_filter_vectors.json
 * (__tests__/namespaceFilter.test.ts); change the algorithm in all of them, or
 * in none.
 *
 * namespace_filter_mode: "all" (default when missing or empty), "include" or
 * "exclude", trimmed and case-insensitive; any other value is an error.
 * namespace_filter_patterns: comma-separated, entries trimmed and empties
 * dropped, at most 100 of at most 128 UTF-8 bytes each. "*" matches any run of
 * characters; everything else is literal; a pattern matches the whole name,
 * case-insensitively. "include" needs a pattern. System names are rejected
 * first, whatever the mode.
 */

export const NAMESPACE_FILTER_MODE_KEY = "namespace_filter_mode"
export const NAMESPACE_FILTER_PATTERNS_KEY = "namespace_filter_patterns"

export type NamespaceFilterMode = "all" | "include" | "exclude"
const MODES: readonly NamespaceFilterMode[] = ["all", "include", "exclude"]

export const MAX_NAMESPACE_PATTERNS = 100
export const MAX_NAMESPACE_PATTERN_BYTES = 128

export const NAMESPACE_NO_MATCH_WARNING = "The namespace filter matched no databases or schemas."

export class NamespaceFilterError extends Error {}

export type NamespaceFilter = {
  mode: NamespaceFilterMode
  patterns: string[]
}

export type NamespaceFilterResult = {
  kept: string[]
  excluded: string[]
  warning: string // "" when there is nothing to warn about
}

// Whole-name match where "*" is the only wildcard. The same iterative
// two-pointer matcher as the Python copy (Go compiles to RE2, which does not
// backtrack): a backtracking regex is exponential in the star count for a name
// that does not match. Both arguments are lower-cased code-point arrays.
function globMatch(pattern: string[], name: string[]): boolean {
  let p = 0
  let n = 0
  let star = -1
  let mark = 0
  while (n < name.length) {
    if (p < pattern.length && pattern[p] !== "*" && pattern[p] === name[n]) {
      p++
      n++
    } else if (p < pattern.length && pattern[p] === "*") {
      star = p
      mark = n
      p++
    } else if (star !== -1) {
      p = star + 1
      mark++
      n = mark
    } else {
      return false
    }
  }
  while (p < pattern.length && pattern[p] === "*") p++
  return p === pattern.length
}

const utf8 = new TextEncoder()

/** Reads the filter from a connection config. Throws NamespaceFilterError. */
export function parseNamespaceFilter(config: Record<string, unknown> | undefined): NamespaceFilter {
  const rawMode = config?.[NAMESPACE_FILTER_MODE_KEY] ?? ""
  const rawPatterns = config?.[NAMESPACE_FILTER_PATTERNS_KEY] ?? ""
  if (typeof rawMode !== "string") throw new NamespaceFilterError("namespace_filter_mode must be a string")
  if (typeof rawPatterns !== "string") {
    throw new NamespaceFilterError("namespace_filter_patterns must be a comma-separated string")
  }
  const mode = (rawMode.trim().toLowerCase() || "all") as NamespaceFilterMode
  if (!MODES.includes(mode)) {
    throw new NamespaceFilterError("namespace_filter_mode must be one of: all, include, exclude")
  }
  const patterns = rawPatterns
    .split(",")
    .map((p) => p.trim())
    .filter((p) => p !== "")
  if (patterns.length > MAX_NAMESPACE_PATTERNS) {
    throw new NamespaceFilterError(`namespace_filter_patterns allows at most ${MAX_NAMESPACE_PATTERNS} patterns`)
  }
  if (patterns.some((p) => utf8.encode(p).length > MAX_NAMESPACE_PATTERN_BYTES)) {
    throw new NamespaceFilterError(
      `each namespace_filter_patterns entry must be at most ${MAX_NAMESPACE_PATTERN_BYTES} bytes`,
    )
  }
  if (mode === "include" && patterns.length === 0) {
    throw new NamespaceFilterError("namespace_filter_mode 'include' needs at least one pattern")
  }
  return { mode, patterns }
}

/** True when the filter can drop a non-system name. */
export function namespaceFilterActive(f: NamespaceFilter): boolean {
  return f.mode === "include" || (f.mode === "exclude" && f.patterns.length > 0)
}

function matchesAny(name: string, f: NamespaceFilter): boolean {
  const lowered = Array.from(name.toLowerCase())
  return f.patterns.some((p) => globMatch(Array.from(p.toLowerCase()), lowered))
}

/** True when name passes: not a system namespace, then the mode applies. */
export function namespaceAllowed(name: string, f: NamespaceFilter, systemNames: readonly string[] = []): boolean {
  const lowered = name.toLowerCase()
  if (systemNames.some((s) => s.toLowerCase() === lowered)) return false
  if (f.mode === "include") return matchesAny(name, f)
  if (f.mode === "exclude") return !matchesAny(name, f)
  return true
}

/**
 * Splits names into kept and excluded, in input order. warning is set when an
 * active filter leaves nothing: a database may not exist yet, so it is not an
 * error.
 */
export function applyNamespaceFilter(
  names: readonly string[],
  f: NamespaceFilter,
  systemNames: readonly string[] = [],
): NamespaceFilterResult {
  const kept: string[] = []
  const excluded: string[] = []
  for (const n of names) (namespaceAllowed(n, f, systemNames) ? kept : excluded).push(n)
  return { kept, excluded, warning: namespaceFilterActive(f) && kept.length === 0 ? NAMESPACE_NO_MATCH_WARNING : "" }
}
